Grid interactivity: Experimenting with drag and drop to set column start and row start

noisysocks opened the pull request:
_**This PR has morphed into more of an exploration of a few different ideas. I'll split the production-ready parts out into separate PRs when ready.**_ 

## What?
In https://github.com/WordPress/gutenberg/pull/59052 I added a new Grid interactivity experiment and the ability to set the column and row span of a grid item using a resizable box.

This PR builds off of that work and adds the ability to explicitly position a grid item by dragging and dropping it to a cell.

## Why?
Part of https://github.com/WordPress/gutenberg/issues/57478.

## How?
Positioning a grid item is done by setting the `style.layout.columnStart` and `style.layout.rowStart` attributes. These are added in https://github.com/WordPress/gutenberg/pull/59483, which this PR is stacked onto.

If a grid item has a `columnStart` and `rowStart` then it is considered _pinned_. This is indicated in the block toolbar using a toggle button. Unpinning a grid item resets `columnStart` and `rowStart` and will result in the grid item once again being positioned automatically by the grid container.

The drag and drop mechanism works by having a `DropZone` in each of the `GridVisualizer`'s cells. There's some trickery here to work around the fact that`GridVisualizer` needs `pointer-events: none` so that you can interact with the blocks that are underneath the visualiser. Pointer events are disabled on the cells until the user begins to drag a block, at which point we enable pointer events so that the drop zones work as expected.

## Testing Instructions
1. Enable the Grid interactivity experiment
2. Insert a Grid block and add some children to it
5. Experiment with drag and drop to set the child's position and using the resize handles to set the child's span

## Screenshots or screencast

https://github.com/WordPress/gutenberg/assets/612155/84a9c5fd-e1a8-4f19-8bac-6a545f329b4d

noisysocks committed:
Use DropZones in GridVisualizer to set positions of grid items

noisysocks committed:
Remove text and icon from drop zone

noisysocks committed:
Don't obscure the in-between block drop zone

noisysocks committed:
Output grid-start styling

noisysocks committed:
Output grid-start styling

noisysocks committed:
Add Column Start and Row Start controls to block inspector

noisysocks committed:
Merge branch 'add/grid-column-and-row-start' into add/grid-interactivity-positioning

noisysocks committed:
Rename grid-visualizer dir to grid-interactivity

noisysocks committed:
Use specificity instead of !important

noisysocks committed:
Add 'Pin to grid' button

noisysocks committed:
Remove weird blank line

noisysocks committed:
Merge remote-tracking branch 'origin/trunk' into add/grid-interactivity-positioning

noisysocks committed:
Remove pin button for now

noisysocks commented:
@SaxonF @tellthemachines I've nixed the concept of "pinning" and the pin button in the toolbar for now. How's this look?

noisysocks committed:
Revert "Remove pin button for now"

This reverts commit bfe3042566b71ad8b3263c8350855e6dca68246f.

ramonjd commented on packages/block-editor/src/components/grid-interactivity/grid-visualizer.js:
```diff
@@ -0,0 +1,169 @@
+
+	return (
+		<BlockPopoverCover
+			className={ classnames( 'block-editor-grid-visualizer', {
```
Somewhere a little `z-index` jig in mobile view is required as it sits over the sidebar

https://github.com/WordPress/gutenberg/assets/6458278/a68b85ee-1889-4a10-be6b-9095cbf11b07

noisysocks replied:
Hm. Why is this happening? When I look in the inspector the visualiser has `z-index = 30` and the list view has `z-index = 10000` 😕

ramonjd replied:
I suspect they're in different stacking contexts? Or a new one needs to be created around the grid visualizer. 

This tool is pretty handy to work it out: https://github.com/gwwar/z-context

tellthemachines replied:
Yeah they're in different stacking contexts but what's causing the issue is the popover content moving to the bottom of the DOM when the viewport is smaller than 782px:

<img width="1848" alt="Screenshot 2024-03-18 at 5 07 37 PM" src="https://github.com/WordPress/gutenberg/assets/8096000/eb8b53a0-af06-4478-991e-8674dd339802">

I'm not sure what the role of `components-popover__fallback-container` is in popover logic, but it appears after the sidebar and is a sibling of `#wpwrap` so I suspect the only way to fix this with CSS would be to assign a non-negative z-index to `#wpwrap`. I doubt that would be free of consequences though 😅

ramonjd commented on packages/block-editor/src/components/grid-interactivity/grid-visualizer.js:
```diff
@@ -0,0 +1,169 @@
+
+function GridVisualizerGrid( { clientId, blockElement } ) {
+	const [ gridInfo, setGridInfo ] = useState( () =>
+		getGridInfo( blockElement )
```
For the new kid: this is so the visualizer matches the underlying grid when resizing the window? Works really well. Nice 🤟🏻 

Feel free to ignore this opinion: "Info" sounds to me like something I'd ask for at a truck stop on my way to an obscure holiday township with no mobile reception.

Just jokes. What about 'properties'? E.g.,  `getBlockElementGridProperties`

noisysocks replied:
It _does_ handle window resizing but that's secondary. The primary purpose is so that we recalculate the grid lines when the grid or its children change size e.g. if the user changes the number of columns in the inspector. 

I didn't want to use the words _attributes_ or _properties_ because I didn't want to imply that we're returning HTML attributes, React props, or CSS properties. I think a generic word like _data_ or _info_ is right. But idk happy to do what you say here 🤷‍♂️

ramonjd replied:
All good! Just a passing comment with as much value as it cost in the first place = `$0` 😄

noisysocks replied:
Good to ask clarifying questions, it means they're answered for future readers too 🙂

ramonjd commented on packages/block-editor/src/components/grid-interactivity/grid-item-pin-toolbar-item.js:
```diff
@@ -0,0 +1,66 @@
+		} );
+	}
+
+	function pinBlock() {
```
I'm out of the loop so ready to be schooled on this, but, what is the value of the "pinned" state? It's not exactly clear to me.

Is it to _indicate_ that a block has a col/row value? Or, to the user, that a block resides in a col/row that is not the start?


https://github.com/WordPress/gutenberg/assets/6458278/4a072850-e712-4866-9775-fd114baf150a

If so, should he pinned state be reset if I drag my block back to the first cell?

noisysocks replied:
The hypothesis is that it would be helpful to distinguish between items that are arranged automatically by the browser versus items that are going to appear in a specific cell because the user set it that way. Otherwise when other content is added the user might not be clear why some blocks move around but other blocks don't.

So we thought we'd try "pinned" as a metaphor for denoting which blocks won't automatically move around. I suspect this might be utterly confusing to users and that we'll have to come up with a better metaphor or interface. But good to get feedback first.

noisysocks replied:
Thinking out loud. (I'm in a typing mood.) A direction I'd like to explore is:

- when the grid is in Auto mode, all grid items are automatically placed by the browser.
- when the grid is in Manual mode, all grid items are manually placed by the user.

It's tricky though because there's lots of ways to add a block to a container that don't involve indicating where you want the block to go, e.g. the default block appender, the inserter, the _Move to_ command, selecting a bunch of blocks and transforming them to a Grid, toggling the Grid block between modes, etc. Do we disable the block appender and inserter for manual Grids and force users to select a cell? Much to think about.

For now I think pinned / unpinned is okay. Things can be clunky while behind the experimental flag.

tellthemachines replied:
What benefit would having to place items in the manual grid manually bring? I think it's nice to have the ability to do that (e.g. create a 6x6 empty grid and populate it with blocks in whatever pattern) but not sure there's any advantage to _forcing_ that behaviour in manual mode.

noisysocks replied:
I'm basically just interested in seeing if creating a really strong distinction between the two modes means that the feature as a whole is more understandable by non technical users. It's something we'll have to test in the field though, I don't think we have the necessary data to decide right now 😀

ramonjd reviewed (commented):
First of all, the drag/drop and resize functionality is working like butter for me. Incredible work.

Since it's behind the experimental flag, take all my comments with a bag-load of Himalayan salt.

noisysocks committed:
Make drop zone 8x8 at minimum

noisysocks committed:
Improve handling of resizing onto empty cells

noisysocks commented:
I'm working on two issues @SaxonF identified:

- [ ] When you drag a block that has a span to the edge, empty columns are created which means that the Grid block has more columns than are set in the block settings.

   ![image_720](https://github.com/WordPress/gutenberg/assets/612155/458018b4-5d73-44dc-8740-f6c962914e02)

   I think the best way to fix this is to now allow blocks to be moved into a position where they overflow the grid. That is, set a `max` on the start input and disallow the drop zone when `start > num columns - span`. 

- [x] Empty cells need a minimum width/height otherwise they're impossible to drag onto.

   ![image_720](https://github.com/WordPress/gutenberg/assets/612155/90b7a05d-128c-4ab7-897c-7de4e19b8cdd)

   I've fixed this by adding a `min-width` and `min-height` to the drop area which can overflow the cell. That is, the cell is still 0x0 but the drop area will be at least 8x8. I've also improved the resize logic to better handle empty cells by taking into account whether we're dragging past the leading or trailing edge of a grid track.

tellthemachines commented on packages/block-editor/src/components/grid-interactivity/grid-item-pin-toolbar-item.js:
```diff
@@ -0,0 +1,66 @@
+			<ToolbarGroup>
+				<ToolbarButton
+					icon={ pinIcon }
+					label={ __( 'Pin to grid' ) }
```
Could we check the state and make this text "Pinned to grid" when it's enabled? Otherwise it's confusing to work out which state it's actually in in text labels mode:

<img width="107" alt="Screenshot 2024-03-15 at 3 16 13 PM" src="https://github.com/WordPress/gutenberg/assets/8096000/55d8d1e4-203f-4b8a-a314-46e1eae6d5c5">

tellthemachines commented on packages/block-editor/src/components/grid-interactivity/grid-item-pin-toolbar-item.js:
```diff
@@ -0,0 +1,66 @@
+		return null;
+	}
+
+	const isPinned = !! layout?.columnStart && !! layout?.rowStart;
```
The manual controls allow setting only one of these so this should probably be an `||`.

tellthemachines commented on packages/block-editor/src/layouts/grid.js:
```diff
@@ -120,11 +120,13 @@ export default {
 				`grid-template-columns: repeat(auto-fill, minmax(min(${ minimumColumnWidth }, 100%), 1fr))`,
-				`container-type: inline-size`
+				'container-type: inline-size'
+				// 'grid-auto-rows: minmax(8px, auto)'
```
these can probably be deleted lol

tellthemachines reviewed (commented):
This is generally working pretty well!

I'm seeing some weirdness with the block movers when moving around pinned blocks:

https://github.com/WordPress/gutenberg/assets/8096000/d6341a0d-6276-46a9-8d4b-de74f6c71fbe

In this case, the block being moved isn't pinned. The second block in the first row is pinned, and it takes a couple clicks to get past it. But then when the block reaches the end of the second row, it can't move to the third and the button actually greys out as if there were no more blocks in the grid (the first block in the third row is also pinned). The grid is set to manual though I'm not sure that makes a difference.

I think the "move right" button greys out if the next slot in the grid is empty:

<img width="1482" alt="Screenshot 2024-03-15 at 3 50 00 PM" src="https://github.com/WordPress/gutenberg/assets/8096000/53869acf-73ba-42d8-bded-8b8c5a809471">

Ideally, the block movers would behave consistently with drag and drop (especially because there's still no way to use drag and drop on smaller screens or with top toolbar enabled).


Note to self: In testing this I'm seeing places where the responsive logic might need fine-tuning but I think it's fine to  to look into it separately (I'm happy to do that 😅)

noisysocks committed:
Don't allow blocks with a >1 span to be dragged too close to the edge

noisysocks committed:
Delete accidental comment

noisysocks committed:
Treat a block as pinned if it has a column start **or** a row start

noisysocks committed:
Adjust toolbar button label if pinned

noisysocks commented:
> I'm seeing some weirdness with the block movers when moving around pinned blocks:

I _think_ what you're seeing is expected. Right now the block movers control where the block appears in the DOM and this doesn't necessarily relate to where the block appears visually.

It helps to open the List View. Here's my attempt at recreating your video but with List View open. You'll see that the movers change the order of the block and not much else.

https://github.com/WordPress/gutenberg/assets/612155/4771d906-9528-47c7-a0c1-082db58491e3

> Ideally, the block movers would behave consistently with drag and drop (especially because there's still no way to use drag and drop on smaller screens or with top toolbar enabled).

Yeah this is the crux of the issue. Right now the block movers control only the DOM order and the drag and drop interaction controls only `columnStart` and `rowStart`. This results in poor accessibility and a confusing experience as you've illustrated.

My preference is to wear this as a known issue for now and then work on improving the situation in a follow-up PR. I can note it as a task in [the overview issue](https://github.com/WordPress/gutenberg/issues/57478). Does that sound okay to you?

> Note to self: In testing this I'm seeing places where the responsive logic might need fine-tuning but I think it's fine to to look into it separately (I'm happy to do that 😅)

Please do!

noisysocks commented:
Thanks for the feedback btw. I've addressed all comments aside from [this z-index one](https://github.com/WordPress/gutenberg/pull/59490#discussion_r1525648914) which I can't (yet) figure out.

noisysocks committed:
Fix visualizer not appearing when block inspector is closed

noisysocks commented:
@SaxonF I improved the treatment of dragging blocks that have a span. Now all of the cells underneath the block are highlighted and you aren't allowed to drop the block too close to the edge of the grid. Give the branch a spin and let me know how you think it's feeling!

https://github.com/WordPress/gutenberg/assets/612155/5d1996ee-b056-4506-aed0-1fa104581bcd

SaxonF commented:
Getting better with every iteration! I'm not seeing the blue drop indicators when testing but the behaviour is working as expected. It prevents moving into cells that cause overlap.

A few details I'd expect:
- I should be able to drag a cell in-between other cells without it pinning. If you don't drop the cell right in the middle of the gap it will move it to selected cell instead, even when there is an indicator suggesting you're dropping inbetween.
- I should be able to drag and a cell before the first cell in a row and after the last cell without pinning
- If I drop a cell back into the row/column it originated from, it shouldn't trigger pinning
- The order of blocks in the list view should represent how they sit on top of each other. At the moment it looks like we are adding z-index on some blocks

<img width="388" alt="image" src="https://github.com/WordPress/gutenberg/assets/1072756/85660591-948c-470a-aa0e-c50eaf3ffc79">

tellthemachines commented on packages/block-editor/src/components/child-layout-control/index.js:
```diff
@@ -223,7 +223,11 @@ export default function ChildLayoutControl( {
-									max={ parentLayout?.columnCount }
+									max={
+										parentLayout?.columnCount -
+										columnSpan +
```
I'm seeing a `Warning: Received NaN for the `max` attribute.` which I think is coming from here; `columnCount` doesn't always exist so this might be NaN on auto layouts.

tellthemachines replied:
Still seeing this error here on Auto layouts.

tellthemachines replied:
Correction: I also see it in Manual mode when clicking "Pin to grid" without moving the block.

noisysocks replied:
Fixed. (I think.)

tellthemachines reviewed (commented):
Hmm the blue highlight on the grid area to drop in has disappeared, is that expected? I'm no longer sure how to move a pinned block to a different location. 

Also, dragging and dropping over the in-between inserter seems to have no effect for a pinned block:

https://github.com/WordPress/gutenberg/assets/8096000/a0558623-08cc-4f2d-960b-6d9589ce1e33

noisysocks committed:
Fix dragging in Chrome

noisysocks commented:
Thanks @SaxonF! Good feedback.

> I'm not seeing the blue drop indicators when testing but the behaviour is working as expected. It prevents moving into cells that cause overlap.

This should be fixed now. Turns out `event.dataTransfer` works differently in Chrome and Firefox.

tellthemachines commented:
> I should be able to drag a cell in-between other cells without it pinning. If you don't drop the cell right in the middle of the gap it will move it to selected cell instead, even when there is an indicator suggesting you're dropping inbetween

What about dragging a pinned block? It feels wrong to rearrange blocks in list view and see nothing happening:


https://github.com/WordPress/gutenberg/assets/8096000/5eb21148-2f05-42d2-add8-3c6686a50ed0


Could we perhaps unpin a block if it's dragged and dropped to a different location?

SaxonF commented:
> Could we perhaps unpin a block if it's dragged and dropped to a different location?

I see this affecting z-index though e.g. you have two overlapping cells and want one to be on top. I do think we should unpin if you drag a pinned item in-between two others on canvas though, so almost the opposite behaviour when dropping on a cell.

paaljoachim commented:
This might be the wrong place to bring it up.... but I am thinking can one also create an overlap of blocks? As in two layer /multi layer approach that partly or fully overlaps each other. So one can for instance add a paragraph block/heading block that crosses on top  of two image blocks etc.

SaxonF commented:
@paaljoachim you can do that with this branch yes

jasmussen commented:
Nice work this is cool. Here's a GIF of me taking this for a bit of a spin:

![state](https://github.com/WordPress/gutenberg/assets/1204802/4dde3f16-d1b6-4c66-a448-0470e9d1bfaa)

A few thoughts describing the above but in no particular order.

The fact that this visualizes the DOM is an interesting challenge. I.e. if I add paragraphs and press Enter, I horizontally populate cells in the grid. If I do this enough, I make a new _row_, because the grid starts wrapping. That allows me to resize vertically cell number 1 to have a row-span of two, and that persists even if at a later time I delete subsequent cells. That's accurate enough to the grid spec, but it does surface some of the challenges around how those resize handles work.

For example, and as is also shown in the GIF, if there is only 1 row, then using the vertical resize handle doesn't do anything. It isn't clear to me whether one of the following two solutions to that would be better:

1. Resizing vertically should always add new rows, if they aren't already there, in a snapping fashion.
2. Or: the vertical resize should only be visible if there's a row to snap into. 

In both cases, a "snapping" behavior would be nice. I believe @talldan worked on something adjacent to this a while back, there may be some learnings there. In any case, dragging a resize handle is fluid and suggests free-form resizing, which isn't the fact. If we have snap points so that a cell snaps into the nearest column or row, it would help address this. 

Resizing from 2 cells down to 1 cell does so in a "centered" fashion, which seems buggy. Ideally the left anchorpoint for the cell resizing remains fixed. 

If I scale a single cell downwards, it also does the centering thing, I also got it to a state where even though I stopped dragging, the resize handle still stuck to my cursor and kept resizing. That's shown towards the end of the GIF.

Probably a separate thing, but made more visible due to the grid sizes here, if I resize a horizontal handle and my cursor overlaps the bottom handle, the bottom hande hover state kicks in. It would be a nice enhancement that the resize handle hover state only ever kicks in, if you're not dragging another handle at the same time. 

Should I be able to resize leftwards, or upwards? In edge or corner cells, no. But in a 9-grid, it seems potentially valid to have all 4 handles available for the center cell. 

That might make it tricky to just click the content inside though, as the resize handles are fairly big, but that's already mitigated a bit by the handles only appearing after you select. But we could also make the handles a bit smaller, potentially, to improve things. After all, these handles are not just the white circle dot but the full width or height of a cell too.

Working in the site editor makes the grid block hard to use because the "Block" tab in the inspector isn't automatically selected when clicking a block.

Multi select across cells currently selects the paragraphs inside. This is probably good to keep, as those are the blockos that actually make the cells appear in the first place. It's also one of those aspects of CSS grid that's really hard to visualize for people that aren't webdesigners. They might see the cells and assume table-like behaviors from Excel of being able to select 4 cells and unify them into one. 

Nice work!

jasmussen commented:
Adding some additional thoughts on the pinning feature:

![pin icon](https://github.com/WordPress/gutenberg/assets/1204802/eec2cdce-d223-4101-a2fe-c4d4ba7e7bab)

Nitpick, we need a better pin icon. I'm happy to help there. Also, the padding in this toolbar group is incorrect if you compare with other toolbar groups. Looks like there are a couple of nested groups going on:

![toolbar group](https://github.com/WordPress/gutenberg/assets/1204802/2363ad64-6721-4d44-8e0b-835f3da8261e)

The feature itself highlights even more, how hard it is to make CSS grid intuitive. But I'd tend to think what the pinning feature is meant to accomplish here should still be allowed; it's using vanilla CSS, and allowing overlap is a valid design tool. 

But it's tricky if you want to edit the text of a lower layer at later time, short of unpinning, and editing. One thing we're doing with #60347 is _bringing the block to the front when selected_. That other PR discusses negative margins, and uses this CSS to accomplish it:

```
	&.is-selected,
	&.has-child-selected {
		// Bring the selected block forward so we can see it.
		z-index: z-index(".block-editor-block-list__block.is-selected");
	}
```

It's not clear that z-index would work the same way for grid, but it's worth a shot, because even an overlapped block you could select using the list view:

![pinning](https://github.com/WordPress/gutenberg/assets/1204802/3a085cfd-4861-42ac-9aa7-dfc1e96f0af2)

Because it allows overlap, I wonder if we should move the toggle from the block toolbar and into the inspector, however. Firstly, that was where I was looking for it, since that's where col/row span aspects are handled. Secondly, it might afford help text, perhaps even being hidden in the ellipsis menu if need be. Alternately, it can exist in both places if we get it to an intuitive enough place — that's always an upgrade you can do at a later time if you start simple.

One more thing, these dashed outlines:
![clicking the dashed outlines](https://github.com/WordPress/gutenberg/assets/1204802/3a6f37a6-193d-4c8d-9dd2-d0134a4d0a7a)

These seem a useful way to highlight notably the gutter between columns and rows. I do wonder if people expect to be able to click such as cell to _add a block there_, which they wouldn't be able to. Perhaps a different visual style for these, a dot-based grid instead? A solid gray (semi-transparent black) background?  That might both add character, and make it feel more like a grid indication than clickable areas.

noisysocks committed:
Merge remote-tracking branch 'origin/trunk' into add/grid-interactivity-positioning

noisysocks committed:
Unpin when re-ordering blocks in the block list

tellthemachines commented:
> I do wonder if people expect to be able to click such as cell to _add a block there_, which they wouldn't be able to.

Would it be good to have this as a feature? Perhaps only on manual grids? Once a bunch of columns and rows have been defined, it would be nice to be able to click on any place in the grid and add a block there, instead of having to add a block next to the existing blocks and then move it into the desired position.

SaxonF commented:
Yep we definitely want that, as well as dragging across cells to create

paaljoachim commented:
The first initial go with testing this out.

I have tried it out using https://playground.wordpress.net/gutenberg.html and was right away confused. I expected to by default see the cells/columns that I could create. I noticed I had to go to manual. Then I could manually add in the amount of cells. But nothing happened. I did not see any additional cells. I expected that when the Grid block was selected that I would see all the cells/columns at the same time showing me their borders within the block border.

1- First area I tripped up.
I did not know how to get going as I did not know how to use the block. The UI on the right and what I saw did not give any obvious hints as how to use it.

2- Not seeing the cells/columns even though it by default says 3. These are just not visible when the Grid block is selected. I expected to see the outline of the 3 columns but it did not show me this.

3- No hints as to what by default is happening. I expected that there would be visual hints but could not locate any.

tellthemachines commented on packages/block-editor/src/components/use-on-block-drop/index.js:
```diff
@@ -336,14 +336,40 @@ export default function useOnBlockDrop(
 
 	const moveBlocks = useCallback(
 		( sourceClientIds, sourceRootClientId, insertIndex ) => {
+			const sourceBlocks = getBlocksByClientId( sourceClientIds );
```
I can't seem to get this code to run at all. Trying to drop between blocks is now causing an `Uncaught TypeError: Cannot convert undefined or null to object` that freezes the whole editor 😬

noisysocks replied:
This should be fixed now.

tellthemachines reviewed (commented):
I'm finding the drag and drop experience here really confusing. Aside from the error/editor freeze mentioned below, it's easy to get to a place where moving a block has unexpected results because there are pinned blocks in the grid, and there's no immediate way of visualising _which_ blocks are pinned unless they're selected. I think at a minimum it would be nice to flag the pin in list view, but ideally we'd have a color overlay or something that would make it really obvious.

Should we consider perhaps consolidating drag to resize before adding the pinning functionality? One thing that still feels a bit wrong is the absence of top and left drag handles.

We could also approach dragging into place iteratively by only exposing the ability to drop onto a currently empty cell (so we'd basically just be extending regular drag and drop) and leaving the ability to overlap blocks by dropping them over other blocks for later.

What do you all think?

noisysocks commented:
Oops forgot to mark this as a draft on Friday.

My plan is to work on the following:

- Remove row/column start attribute when block is dragged “in between” blocks
  - This is done, but I need to work out a less hacky way of achieving it.
- Disable dragging a block onto an occupied cell
- Only set row/column start attribute if block wouldn’t end up there by simply re-ordering it
- Address feedback

noisysocks committed:
Disable dragging a block onto an occupied cell

noisysocks committed:
Split calculateGridRect

noisysocks committed:
Move the block instead of setting a row/column start when possible

noisysocks committed:
Merge remote-tracking branch 'origin/trunk' into add/grid-interactivity-positioning

noisysocks commented:
OK I've pushed a rough pass at implementing those first three dot points. @SaxonF @tellthemachines let me know how this feels and what bugs you run into while I start working on cleaning it up.

SaxonF commented:
This is starting to feel nice! I think we are honing in on the right direction which is doing our best to guess whether a block is pinned or not based on where its dropped and the position its dropped in, but when needed power users can manually set pinning.

I ran into some bugs along the way:

The "centered" resizing issue Joen highlighted above . Shrinking a column doesn't work a lot of the time.

I should be able to drag and drop the third block here from column 2 to column 3 even though its on top of itself . Relatedly we may want to support dropping on occupied cells in future to allow for overlapping layouts. Starting with this constraint is fine though as we don't have to deal with as many z-index issues.

<img width="1512" alt="image" src="https://github.com/WordPress/gutenberg/assets/1072756/e57ad82f-ef0f-43bf-8f8f-d0b9a4d861c1">

Sometimes the dragging can be a little unreliable . In the below example I would have assumed I could drag the second grid item into the first column of the second row. I suspect this is mostly due to trying to remove pinning if a grid item is placed in its natural position. If the list order doesn't match with the visual order it can be confusing. This will be something we need to keep an eye on during testing.

https://github.com/WordPress/gutenberg/assets/1072756/2f74e5c9-3c10-49df-80e2-503e19830bf6

Assume I have 4 columns and 3 grid items. I move the first grid item into the 4 column cell. I would expect it to remain at the end which means we'd need to also move it in the list order which would be identical to dropping it in-between cells at the end.

- Moving a g

SaxonF commented:
After we get these little issues sorted we should move on to clicking and dragging cells to add blocks. That will give us a really good feel for what this experience can become. Then we can work on tightening up all the visual details and micro interactions.

noisysocks committed:
Rename blockElement -> gridElement

noisysocks committed:
Don't moveBlockToPosition() for now. The algorithm needs to be robust

noisysocks committed:
Take span into account when validating drag

noisysocks commented:
Thanks @SaxonF!

> The "centered" resizing issue Joen highlighted above . Shrinking a column doesn't work a lot of the time.

I can't seem to reproduce this. I checked Firefox and Chrome. Is there some step that you need to do first? 

> I should be able to drag and drop the third block here from column 2 to column 3 even though its on top of itself . Relatedly we may want to support dropping on occupied cells in future to allow for overlapping layouts. Starting with this constraint is fine though as we don't have to deal with as many z-index issues.

I've fixed this. Let's look more into overlapping blocks in a follow-up PR. Perhaps if you are holding down <kbd>Cmd</kbd> then we can disable drag validation and let you drop the block wherever you want.

> Sometimes the dragging can be a little unreliable . In the below example I would have assumed I could drag the second grid item into the first column of the second row. I suspect this is mostly due to trying to remove pinning if a grid item is placed in its natural position. If the list order doesn't match with the visual order it can be confusing. This will be something we need to keep an eye on during testing.

Hmm, yes, bugger. My approach of simply checking whether the leading cell is empty is too naive. This PR is already doing quite a lot so let's remove this for now and handle it in a follow-up PR. As part of this I'll also tackle re-ordering the DOM to match the visual order.

noisysocks commented:
Making my way through all of the feedback. Your turn now @tellthemachines 😀

> I'm finding the drag and drop experience here really confusing. Aside from the error/editor freeze mentioned below, it's easy to get to a place where moving a block has unexpected results because there are pinned blocks in the grid, and there's no immediate way of visualising _which_ blocks are pinned unless they're selected. I think at a minimum it would be nice to flag the pin in list view, but ideally we'd have a color overlay or something that would make it really obvious.

What do you think of these design suggestions @SaxonF? Happy to try them.

> Should we consider perhaps consolidating drag to resize before adding the pinning functionality? One thing that still feels a bit wrong is the absence of top and left drag handles.

There's a bug in `ResizableBox` causing the top and left drag handles to not work properly right now. I'll add this as a task in the overview issue to come back to.

> We could also approach dragging into place iteratively by only exposing the ability to drop onto a currently empty cell (so we'd basically just be extending regular drag and drop) and leaving the ability to overlap blocks by dropping them over other blocks for later.

This is implemented 👍

noisysocks commented:
> I have tried it out using https://playground.wordpress.net/gutenberg.html and was right away confused. I expected to by default see the cells/columns that I could create. I noticed I had to go to manual. Then I could manually add in the amount of cells. But nothing happened. I did not see any additional cells. I expected that when the Grid block was selected that I would see all the cells/columns at the same time showing me their borders within the block border.

@paaljoachim: Did you enable the experiment in Gutenberg → Experiments? The grid outlines won't show unless that setting is turned on.

noisysocks committed:
Don't put pin toolbar button in a group

noisysocks committed:
Add translucent background to grid cells

noisysocks commented:
Belated thanks for the comments @jasmussen.

> For example, and as is also shown in the GIF, if there is only 1 row, then using the vertical resize handle doesn't do anything. It isn't clear to me whether one of the following two solutions to that would be better:

I think this is a lot better now that you can specify the number of rows that you'd like when the grid is in manual mode. I quite like the idea of resizing the grid automatically when you resize an item past the grids bounds, but let's follow up on this later.

> In both cases, a "snapping" behavior would be nice. I believe @talldan worked on something adjacent to this a while back, there may be some learnings there. In any case, dragging a resize handle is fluid and suggests free-form resizing, which isn't the fact. If we have snap points so that a cell snaps into the nearest column or row, it would help address this.

I like this idea. It will require some changes to the public `ResizableBox` component so let's constrain the scope of this PR and circle back to it later. I've noted it in the overview issue.

> Nitpick, we need a better pin icon.

Create one and I'll slap it in!

> Also, the padding in this toolbar group is incorrect if you compare with other toolbar groups.

Fixed.

> Because it allows overlap, I wonder if we should move the toggle from the block toolbar and into the inspector, however. Firstly, that was where I was looking for it, since that's where col/row span aspects are handled. Secondly, it might afford help text, perhaps even being hidden in the ellipsis menu if need be. Alternately, it can exist in both places if we get it to an intuitive enough place — that's always an upgrade you can do at a later time if you start simple.

Happy to try this. Maybe as a follow up? I'd appreciate a mockup of what exactly this would look like with the help text.

> These seem a useful way to highlight notably the gutter between columns and rows. I do wonder if people expect to be able to click such as cell to add a block there, which they wouldn't be able to.

This is coming—it's on the [todo list](https://github.com/WordPress/gutenberg/issues/57478).

> Perhaps a different visual style for these, a dot-based grid instead? A solid gray (semi-transparent black) background? That might both add character, and make it feel more like a grid indication than clickable areas.

I took your suggestion quite literally and made the border `dotted` and added a 10% transparent `$gray-400` background. I think it does look nicer with the background. Please feel free to push to the branch any further tweaks you'd like.

noisysocks commented on packages/block-editor/src/components/use-on-block-drop/index.js:
```diff
@@ -336,14 +336,40 @@ export default function useOnBlockDrop(
 		( sourceClientIds, sourceRootClientId, insertIndex ) => {
+			const sourceBlocks = getBlocksByClientId( sourceClientIds );
+
+			// TODO: Figure out a better place for this. useOnBlockDrop shouldn't care about layout styles.
```
I'm struggling to come up with a better place/approach for this. Any ideas @tellthemachines?

tellthemachines replied:
I wouldn't worry about contaminating drag and drop with layout logic; `useOnBlockDrop` [already cares about layout styles](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/use-on-block-drop/index.js#L291) 😅 

Drag and drop doesn't exist in a vacuum. It's an interaction that's tied to an editable canvas, and it makes sense that its possibilities are determined by layout, block types and block state. 

~~Would it be worth checking that there's a pin in place before running `updateBlockAttributes` on all the blocks being moved?~~ Edit: I see Andy already asked the same question 😄

noisysocks committed:
Fix input max

noisysocks commented:
Marking this as ready again. I've addressed a lot of the feedback and updated the [overview issue](https://github.com/WordPress/gutenberg/issues/57478) with some tasks to follow up on. I'm not intending to land this in a perfect state, just needs to be a rough base that we can build from.

The two blocking items remaining I think are the centre resizing issue that Saxon and Joen have ran into (that I can't reproduce) and figuring out a good solution for https://github.com/WordPress/gutenberg/pull/59490#discussion_r1568085096.

andrewserong commented on packages/block-editor/src/components/use-on-block-drop/index.js:
```diff
@@ -336,14 +336,40 @@ export default function useOnBlockDrop(
+
+			// TODO: Figure out a better place for this. useOnBlockDrop shouldn't care about layout styles.
+			const unpinBlock = () => {
+				updateBlockAttributes(
```
Tiny nit: would it be worth skipping calling `updateBlockAttributes` if no blocks need to be unpinned? I.e. move the `sourceBlocks.reduce` to be before this call?

I suppose this note is similar to your comment about `useOnBlockDrop shouldn't care about layout styles`, but unfortunately I can't think of a better place for this logic to live, either, without awkward `useEffect`s 🤔

noisysocks replied:
`updateBlockAttributes` is _supposed_ to do nothing if you pass it `{}`. (It doesn't, hence that change to the reducer that I have in this PR...)

Probably better to skip the call altogether though, yeah 👍

tellthemachines reviewed (commented):
Playing around with this a bit more, it's much improved with the constraints, but odd things still happen:

* initiating drag and then dropping the block in its previous position pins it
* it's almost impossible to move a block to the last position in the child block list without pinning it
* drop indicators sometimes appear in places they shouldn't, or don't appear when they should (this happens a little on trunk too, but is more pronounced on this branch)
* sometimes trying to drop on an in-between indicator does nothing

I'd also expect that pinning a block to a position after all the other blocks would reorder it in the markup _as well as_ pinning it. This is important to get right for accessibility/usability; if visual order doesn't match markup and list view order it gets really confusing. I don't think we should merge this PR before that at least is solved, otherwise with 18.2 RC coming up this week we risk shipping a broken experience in the next plugin version. I know this is behind the experimental flag but folks have been using it quite a bit regardless 😅 

Edit: here's a video showing the above points:



https://github.com/WordPress/gutenberg/assets/8096000/cde3d03e-6627-43a1-ba85-549986e57770

SaxonF commented:
> and there's no immediate way of visualising which blocks are pinned unless they're selected

We should definitely do this

SaxonF commented:
> I'm finding the drag and drop experience here really confusing. Aside from the error/editor freeze mentioned below, it's easy to get to a place where moving a block has unexpected results because there are pinned blocks in the grid

It occurred to me that some of the confusion stems from the fact that we don't yet show the end result of a drag/drop (like we do in list views for example). You can play with react-grid-layout to get an idea of how this would feel https://react-grid-layout.github.io/react-grid-layout/examples/0-showcase.html. Pinned items would remain stuck in place.

https://github.com/WordPress/gutenberg/assets/1072756/192fa735-7e63-4b76-b9cd-713888c99a2c

tellthemachines commented:
> It occurred to me that some of the confusion stems from the fact that we don't yet show the end result of a drag/drop (like we do in list views for example).

Should we implement that before this grid work then? Also, in that example, the static blocks are visually distinguishable from the others. I'm not sure it would be so clear if they all looked exactly the same.

SaxonF commented:
> I'm not sure it would be so clear if they all looked exactly the same.

Yes we'd want to distinguish between them, even if that's just when starting to drag.

tellthemachines commented:
Oh, one other item worth adding to the list of follow-ups: it should be possible to drag a block from outside of the grid into a pinned position on the grid.

noisysocks commented:
> It occurred to me that some of the confusion stems from the fact that we don't yet show the end result of a drag/drop (like we do in list views for example). You can play with react-grid-layout to get an idea of how this would feel https://react-grid-layout.github.io/react-grid-layout/examples/0-showcase.html. Static blocks in the demo are similar to pinned items.

This is really a drag and drop enhancement, not specific to grid. I don't _think_ we can untangle them. I dug up https://github.com/WordPress/gutenberg/pull/16457 which is an old exploration into doing it that was abandoned because it didn't play nicely with nested blocks. 

Maybe something to think about but probably not a can of worms we want to open right now.

noisysocks commented:
> > and there's no immediate way of visualising which blocks are pinned unless they're selected
> 
> We should definitely do this

How should we visualise it?

noisysocks commented:
> I'd also expect that pinning a block to a position after all the other blocks would reorder it in the markup _as well as_ pinning it. This is important to get right for accessibility/usability; if visual order doesn't match markup and list view order it gets really confusing. I don't think we should merge this PR before that at least is solved

Alright I'll work on a better solution to this. My previous attempt was too naive.

ramonjd commented:
Sorry if this has been covered before - on the face of it I'm finding undo seems to work for some interactions and not others.

https://github.com/WordPress/gutenberg/assets/6458278/debbc440-924b-420e-8be1-260874dd21a5

I think it's tied to the cell. So if I'm editing a single cell the history isn't saved. If I edit different ones it appears to work as expected


https://github.com/WordPress/gutenberg/assets/6458278/87a9bdc5-af4f-453d-870a-4acd39a56a2e

jasmussen commented:
@noisysocks for [this comment](https://github.com/WordPress/gutenberg/pull/59490#issuecomment-2060146839), can you use `pin-small` from [the set](https://wordpress.github.io/gutenberg/?path=/docs/icons-icon--docs)? The regular pin is more about "posts".

noisysocks committed:
(Not fully working:) Call moveToPosition() when moving a block to its 'natural position'

noisysocks commented:
I got halfway through addressing [this comment](https://github.com/WordPress/gutenberg/pull/59490#pullrequestreview-2004839224) in [b70baee](https://github.com/WordPress/gutenberg/pull/59490/commits/b70baeeea1effadce65ed41f37fab7a50a7af4ab). It's pretty complicated to work out when a block needs a `columnStart` and `rowStart` versus when it doesn't. We essentially need to implement some of the [grid auto placement algorithm](https://www.w3.org/TR/css-grid-1/#auto-placement-algo). 

Instead of doing that 😅 I am thinking we should look at adding constraints. For example not allow a user to set `columnStart` without also setting `rowStart` and vice versa.

I'm also curious about creating a strong distinction between auto mode and manual mode where, in manual mode, all grid items have a `columnStart` and `rowStart` and we disable re-ordering blocks altogether. This way there's never any mixing of items that are placed manually and items that are placed automatically. This will simplify things greatly (e.g. no "pinning"), though at the cost of Grid behaving differently to other container blocks (e.g. no block movers). I'll explore this in a separate PR next week some time.

For now though @SaxonF is hounding me about wanting to see how "click or drag a cell to insert a block" feels so I'm going to work on that.

Safe to say this PR is a grabbag of a few different ideas, none of which are production ready, and so I've marked it as a draft again.

noisysocks committed:
Very very rough crack at drag to insert

noisysocks commented:
> For now though @SaxonF is hounding me about wanting to see how "click or drag a cell to insert a block" feels so I'm going to work on that.

I've pushed a barely-usable first pass at this 😀

SaxonF commented:
"hounding" 🤣  . Accurate though!

Agree we should re-explore everything pinned in manual (ie no pinning). When we tried last time we didn't have rows in place plus a couple of the other improvements which affected how it felt. "click or drag a cell to insert a block" will give us a more complete experience to try.

We can frame "type" as grid item position and highlight the difference.

![image](https://github.com/WordPress/gutenberg/assets/1072756/645eb55d-e695-4842-bf3a-0c4d759273c1)

SaxonF commented:
One other note. When dragging a grid item we should add outlines to all grid items (like when you hover over them) so its clear where their boundaries are.

tellthemachines commented:
> For example not allow a user to set `columnStart` without also setting `rowStart` and vice versa.
> 
> I'm also curious about creating a strong distinction between auto mode and manual mode where, in manual mode, all grid items have a `columnStart` and `rowStart` and we disable re-ordering blocks altogether.

If all grid items have a `columnStart` and `rowStart`, ensuring both are set will be much easier because column and row start fields will be pre-populated with the default values 👍 

> This will simplify things greatly (e.g. no "pinning"), though at the cost of Grid behaving differently to other container blocks (e.g. no block movers).

I think if we do this, it's important that the experience be consistent with other blocks, though the underlying behaviour may be different. E.g. we should make sure it's still possible to move blocks in list view, and perhaps replace the default block movers with a two-dimensional set of controls that allows to move both vertically and horizontally (this was @noisysocks' suggestion in conversation with me earlier). 

Ideally we want the action of moving blocks, whether by drag and drop or keyboard, to be the same as it is for all other blocks _except_ when these blocks move within the grid, their `columnStart` and `rowStart` attributes change. 

The only unique visible behaviour is that we want to be able to drag or keyboard move one of these blocks to any empty grid cell (that might not be possible to do in list view, but it should be doable through either the block movers or the column and row start controls, apart from canvas drag and drop).

Grid is a unique form of container in that its configuration determines the exact space its children occupy, and it displays empty grid cells if there aren't enough children to fill it. It's quite different from flex in this respect, because in flex the size of the child to some extent determines the distribution of space, and unless the children are few and very small, there is no extra space within the container. I'm rambling a bit 😅  but what I'm trying to get at is that grid cells are a sort of virtual container that we could treat similarly to how empty paragraphs are treated if we wanted them to have a presence in list view. We're already kind of treating them as such with the ability to drop into any grid cell.

SaxonF commented:
Iterating on the [above](https://github.com/WordPress/gutenberg/pull/59490#issuecomment-2065732213). If we did make it so that in manual mode everything has a position. Based on [this conversation](https://github.com/WordPress/gutenberg/pull/49050#issuecomment-2044142776) an alternative would be adding max column setting on auto mode which would cover most scenarios 

![image](https://github.com/WordPress/gutenberg/assets/1072756/c25c2e6c-835d-4702-95fd-416294702d88)

tellthemachines commented:
> See this example here of a grid that is a little more complex

@fabiankaegy that looks like a bug in the responsive logic for grid elements spanning multiple columns. I'll look into it, thanks for flagging!
